### PR TITLE
two new matrix norm

### PR DESCRIPTION
**Pull request by yya007-zz**

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
 
+Tensor frobenius_norm(const Tensor& self) {
+  IntList dim = {0, 1};
+  return at::native::frobenius_norm(self, dim);
 }
+
+Tensor frobenius_norm(const Tensor& self, IntList dim, bool keepdim) {
+  AT_CHECK(
+      dim.size() <= 2,
+      "Improper number of chosend dimensions to do frobenius norm.");
```
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: I think this message can be improved:
```
"Expected at most 2 dimensions, but got ", dim.size(), " dimensions instead."
```

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
+  if (dim.size() == 0) {
```
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Why would this case arise?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
+    return at::native::norm(self, 2, dim[0], keepdim);
+  }
+  Tensor result = self.type().tensor();
+  return at::sqrt(at::native::sum(self * self, dim, keepdim));
 }
+
+Tensor nuclear_norm(const Tensor& self) {
+  AT_CHECK(
+      self.dim() == 2,
+      "Improper number of tensor dimensions to do nuclear norm.");
```
Comment (vishwakftw) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Same as above.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1741,6 +1741,15 @@
   python_default_init:
     p: 2
 
+- func: frobenius_norm(Tensor self) -> Tensor
+  variants: method, function
```
Comment (vishwakftw) [inline aten/src/ATen/native/native_functions.yaml]: `variants` are `method` and `function` by default. I don't think it's necessary to specify them explicitly.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -2971,6 +2971,72 @@ def parse_kwargs(desc):
...
 """)
 
+add_docstr(torch.frobenius_norm,
+           r"""
+.. function:: frobenius_norm(input) -> Tensor
+
+Returns the frobenius norm of the :attr:`input` tensor.
+
+.. math::
+    ||x||_{F} = \sqrt{tr(xx^{H})}
```
Comment (vishwakftw) [inline torch/_torch_docs.py]: Could you please change `H` to `T`. We don't have support for complex tensors yet.

**Review by vishwakftw — COMMENTED** [review]
ONNX changes have been added as a part of the commit. Is this intentional?

Could you fix the lint errors too?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (yya007) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Dim are the dimensions will be reduced. You could not reduce more than two dimensions when calculating Frobenius Norm.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (yya007) [reply aten/src/ATen/native/LinearAlgebra.cpp]: As the title of this PR, this is a matrix norm and we want to restrict the input to be a 2D tensor.

Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Oh no, sorry about the confusion. I meant that the error message can be improved.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: I am guessing the parsing happens like this:
```
frobenius_norm(torch.Tensor, python list, python bool)
```
Why would someone pass an empty list?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 6]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
+Tensor frobenius_norm(const Tensor& self, IntList dim, bool keepdim) {
+  AT_CHECK(
+      dim.size() <= 2,
+      "Improper number of chosend dimensions to do frobenius norm.");
+  if (dim.size() == 0) {
+    Tensor result = self.type().tensor();
+    return at::native::norm_out(result, self, 2, self.dim(), keepdim);
+  }
+  if (dim.size() == 1) {
+    return at::native::norm(self, 2, dim[0], keepdim);
```
Comment (SsnL) [inline aten/src/ATen/native/LinearAlgebra.cpp]: All these native calls break autograd. Please use the top namespace ones, e.g., `at::norm`

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 7]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
+  return at::native::frobenius_norm(self, dim);
 }
+
+Tensor frobenius_norm(const Tensor& self, IntList dim, bool keepdim) {
+  AT_CHECK(
+      dim.size() <= 2,
+      "Improper number of chosend dimensions to do frobenius norm.");
+  if (dim.size() == 0) {
+    Tensor result = self.type().tensor();
+    return at::native::norm_out(result, self, 2, self.dim(), keepdim);
```
Comment (SsnL) [inline aten/src/ATen/native/LinearAlgebra.cpp]: autograd doesn't support `_out` variants. You shout just use `norm` here.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 8]
```diff
@@ -315,5 +315,32 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
...
+  }
+  if (dim.size() == 1) {
+    return at::native::norm(self, 2, dim[0], keepdim);
+  }
+  Tensor result = self.type().tensor();
+  return at::sqrt(at::native::sum(self * self, dim, keepdim));
 }
+
+Tensor nuclear_norm(const Tensor& self) {
+  AT_CHECK(
```
Comment (SsnL) [inline aten/src/ATen/native/LinearAlgebra.cpp]: If we support batch frobenius norm, why do we not support nuclear norm? After all, they should eventually will be called from the same python function.

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -2971,6 +2971,72 @@ def parse_kwargs(desc):
...
+
+Example::
+
+    >>> a = torch.randn(2, 3)
+    >>> a
+    tensor([[-0.6971,  1.1057, -0.3520],
+            [-2.1726, -1.0937,  0.6892]])
+    >>> a.frobenius_norm((0,1))
+    tensor(2.8677)
+    >>> a.frobenius_norm((0,1),True)
```
Comment (SsnL) [inline torch/_torch_docs.py]: space before True

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -2971,6 +2971,72 @@ def parse_kwargs(desc):
...
+Example::
+
+    >>> a = torch.randn(2, 3)
+    >>> a
+    tensor([[ 1.0043,  0.5987, -1.2026],
+            [ 1.7241, -2.6561,  0.8088]])
+    >>> a.frobenius_norm()
+    tensor(3.6735)
+
+.. function:: frobenius_norm(input, dim, keepdim=False) -> Tensor
```
Comment (SsnL) [inline torch/_torch_docs.py]: I'd rather dispatch to theses functions in `norm` (i.e., what matlab and numpy do) rather than document them separately.

**Review by ssnl — COMMENTED** [review]
Didn't check the math. 

Also please revert the submodule change

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 8 above)
Comment (vishwakftw) [reply aten/src/ATen/native/LinearAlgebra.cpp]: There is one caveat, `gesvd` is not implemented for batched tensors. But I agree, we must have something like `numpy.linalg.norm()` for all of this.

Comment (yya007) [reply aten/src/ATen/native/LinearAlgebra.cpp]: The svd function don't support batched tensors now. This PR intends to create the matrix norms only. Extending the function to batched matrix norm is what we want to do in the future.  We may have another PR to let the nuclear norm to support batched norm.

**Comment by yya007** [comment]
Wrong merge. Abandon and resubmit.

**Pull request closed without merging**